File: sdk/typescript/src/execution/engine.ts
/**
 * @centris/sdk - Execution Engine
 *
 * Executes capabilities using the planned method with retry and fallback support.
 */

import type {
  ExecutionPlan,
  ExecutionMethod,
  ExecutionOptions,
  ExecutorContext,
  ExecutorResult,
  Executor,
} from "./types.js";
import { APIExecutor } from "./executors/api.js";
import { BrowserExecutor } from "./executors/browser.js";
import { DesktopExecutor } from "./executors/desktop.js";
import { ExecutionRouter, createExecutionRouter } from "./router.js";
import { executorError } from "./types.js";

/**
 * Execution Engine
 *
 * Executes capabilities according to execution plans, handling:
 * - Primary method execution
 * - Automatic fallbacks on failure
 * - Retries with exponential backoff
 * - Confirmation requests for sensitive actions
 */
export class ExecutionEngine {
  private router: ExecutionRouter;
  private executors: Map<ExecutionMethod, Executor>;
  private confirmationHandler?: (plan: ExecutionPlan) => Promise<boolean>;

  constructor(options?: {
    router?: ExecutionRouter;
    confirmationHandler?: (plan: ExecutionPlan) => Promise<boolean>;
  }) {
    this.router = options?.router ?? createExecutionRouter();
    this.confirmationHandler = options?.confirmationHandler;

    // Initialize executors
    this.executors = new Map();
    this.executors.set("api", new APIExecutor());
    this.executors.set("browser", new BrowserExecutor());
    this.executors.set("desktop", new DesktopExecutor());
  }

  /**
   * Execute a capability with automatic routing.
   */
  async execute(params: {
    connectorId: string;
    capabilityId: string;
    params: Record<string, unknown>;
    context: ExecutorContext;
    options?: ExecutionOptions;
  }): Promise<ExecutorResult> {
    // Create execution plan
    const plan = await this.router.planExecution({
      connectorId: params.connectorId,
      capabilityId: params.capabilityId,
      params: params.params,
      context: params.context,
    });

    // Execute according to plan
    return this.executePlan(plan, params.context, params.options);
  }

  /**
   * Execute according to an execution plan.
   */
  async executePlan(
    plan: ExecutionPlan,
    context: ExecutorContext,
    options?: ExecutionOptions,
  ): Promise<ExecutorResult> {
    // Request confirmation if needed
    if (plan.requiresConfirmation) {
      const confirmed = await this.requestConfirmation(plan);
      if (!confirmed) {
        return executorError("CANCELLED", "User declined confirmation", { retryable: false });
      }
    }

    const maxRetries = options?.retries ?? 2;
    const timeout = options?.timeout ?? 30000;
    const methodsToTry = [plan.method, ...plan.fallbackMethods];

    // Try each method
    for (const method of methodsToTry) {
      const executor = this.executors.get(method);
      if (!executor) {
        continue;
      }

      // Check if executor is available
      const available = await executor.isAvailable();
      if (!available) {
        continue;
      }

      // Try with retries
      for (let attempt = 0; attempt <= maxRetries; attempt++) {
        try {
          const result = await this.executeWithTimeout(
            executor,
            plan.connectorId,
            plan.capabilityId,
            plan.params,
            context,
            timeout,
          );

          if (result.ok) {
            return result;
          }

          // Check if error is retryable
          if (!result.error.retryable) {
            break; // Move to fallback
          }

          // Exponential backoff before retry
          if (attempt < maxRetries) {
            await this.delay(Math.pow(2, attempt) * 100);
          }
        } catch (err) {
          // Unexpected error, try fallback
          console.error(`Execution failed (${method}):`, err);
          break;
        }
      }
    }

    // All methods failed
    return executorError(
      "ALL_METHODS_FAILED",
      `All execution methods failed for ${plan.connectorId}.${plan.capabilityId}`,
      { retryable: false },
    );
  }

  /**
   * Execute with timeout.
   */
  private async executeWithTimeout(
    executor: Executor,
    connectorId: string,
    capabilityId: string,
    params: Record<string, unknown>,
    context: ExecutorContext,
    timeout: number,
  ): Promise<ExecutorResult> {
    return new Promise((resolve, reject) => {
      const timer = setTimeout(() => {
        resolve(
          executorError("TIMEOUT", `Execution timed out after ${timeout}ms`, { retryable: true }),
        );
      }, timeout);

      executor
        .execute(connectorId, capabilityId, params, context)
        .then((result) => {
          clearTimeout(timer);
          resolve(result);
        })
        .catch((err) => {
          clearTimeout(timer);
          reject(err);
        });
    });
  }

  /**
   * Request user confirmation for sensitive actions.
   */
  private async requestConfirmation(plan: ExecutionPlan): Promise<boolean> {
    if (this.confirmationHandler) {
      return this.confirmationHandler(plan);
    }
    // Default: auto-confirm (would be overridden in UI)
    return true;
  }

  /**
   * Delay helper for retries.
   */
  private delay(ms: number): Promise<void> {
    return new Promise((resolve) => setTimeout(resolve, ms));
  }

  /**
   * Register a custom executor.
   */
  registerExecutor(method: ExecutionMethod, executor: Executor): void {
    this.executors.set(method, executor);
  }

  /**
   * Set confirmation handler.
   */
  setConfirmationHandler(handler: (plan: ExecutionPlan) => Promise<boolean>): void {
    this.confirmationHandler = handler;
  }

  /**
   * Get the router instance.
   */
  getRouter(): ExecutionRouter {
    return this.router;
  }
}

/**
 * Create an execution engine instance.
 */
export function createExecutionEngine(options?: {
  router?: ExecutionRouter;
  confirmationHandler?: (plan: ExecutionPlan) => Promise<boolean>;
}): ExecutionEngine {
  return new ExecutionEngine(options);
}

File: sdk/python/centris_sdk/execution/engine.py
"""
Centris SDK Execution Engine

The main execution engine that coordinates executors and handles
retry logic, fallbacks, and error handling.
"""

import asyncio
import logging
import time
from typing import Optional, Any

from centris_sdk.types import ExecutionMethod
from centris_sdk.execution.types import (
    ExecutionConfig,
    ExecutionRequest,
    ExecutionResponse,
    Executor,
)
from centris_sdk.execution.router import ExecutionRouter
from centris_sdk.execution.executors.api import APIExecutor
from centris_sdk.execution.executors.browser import BrowserExecutor
from centris_sdk.execution.executors.desktop import DesktopExecutor


logger = logging.getLogger("centris.execution.engine")


class ExecutionEngine:
    """
    Main execution engine for Centris connectors.
    
    Features:
    - Multi-method execution (API, Browser, Desktop)
    - Automatic method selection
    - Retry with exponential backoff
    - Fallback to alternative methods
    - Error handling and recovery
    
    Example:
        engine = ExecutionEngine()
        
        result = await engine.execute(
            connector_id="slack",
            capability_id="send_message",
            params={"channel": "#general", "message": "Hello"},
            context={"access_token": "xoxb-..."},
        )
        
        if result.success:
            print(f"Sent message: {result.data}")
        else:
            print(f"Error: {result.error}")
    """
    
    def __init__(
        self,
        config: Optional[ExecutionConfig] = None,
        auto_setup: bool = True,
    ):
        self.config = config or ExecutionConfig()
        self.router = ExecutionRouter(self.config)
        self._setup_complete = False
        
        if auto_setup:
            # Register default executors
            self.router.register_executor(APIExecutor(self.config))
            self.router.register_executor(BrowserExecutor(self.config))
            self.router.register_executor(DesktopExecutor(self.config))
    
    def register_executor(self, executor: Executor) -> None:
        """Register a custom executor."""
        self.router.register_executor(executor)
    
    async def setup(self) -> None:
        """Initialize all executors."""
        if self._setup_complete:
            return
        
        for method in self.router.get_available_methods():
            executor = self.router.get_executor(method)
            if executor:
                try:
                    await executor.setup()
                except Exception as e:
                    logger.warning(f"Failed to setup {method.value} executor: {e}")
        
        self._setup_complete = True
    
    async def teardown(self) -> None:
        """Clean up all executors."""
        for method in self.router.get_available_methods():
            executor = self.router.get_executor(method)
            if executor:
                try:
                    await executor.teardown()
                except Exception as e:
                    logger.warning(f"Failed to teardown {method.value} executor: {e}")
        
        self._setup_complete = False
    
    async def execute(
        self,
        connector_id: str,
        capability_id: str,
        params: dict[str, Any],
        context: Optional[dict[str, Any]] = None,
        capability: Optional[Any] = None,
        preferred_method: Optional[ExecutionMethod] = None,
        timeout: Optional[float] = None,
    ) -> ExecutionResponse:
        """
        Execute a capability.
        
        Args:
            connector_id: ID of the connector
            capability_id: ID of the capability to execute
            params: Input parameters
            context: Execution context (auth tokens, etc.)
            capability: Optional capability reference for method hints
            preferred_method: Preferred execution method
            timeout: Custom timeout
            
        Returns:
            ExecutionResponse with result or error
        """
        await self.setup()
        
        # Build request
        request = ExecutionRequest(
            connector_id=connector_id,
            capability_id=capability_id,
            params=params,
            context=context or {},
            preferred_method=preferred_method,
            timeout=timeout or self.config.default_timeout,
            capability=capability,
        )
        
        # Get capability methods if available
        capability_methods = None
        if capability and hasattr(capability, "execution_methods"):
            capability_methods = capability.execution_methods
        
        # Select executor
        executor = await self.router.select_executor(request, capability_methods)
        if not executor:
            return ExecutionResponse(
                success=False,
                error="No available executor for this capability",
                error_code="NO_EXECUTOR",
                request_id=request.request_id,
            )
        
        # Execute with retry
        return await self._execute_with_retry(
            request,
            executor,
            capability_methods,
        )
    
    async def _execute_with_retry(
        self,
        request: ExecutionRequest,
        executor: Executor,
        capability_methods: Optional[list[ExecutionMethod]],
    ) -> ExecutionResponse:
        """Execute with retry and fallback logic."""
        last_error: Optional[str] = None
        retry_count = 0
        fallback_used = False
        
        current_executor = executor
        tried_methods: set[ExecutionMethod] = set()
        
        while retry_count <= self.config.max_retries:
            tried_methods.add(current_executor.method)
            
            start_time = time.time()
            
            try:
                # Execute with timeout
                response = await asyncio.wait_for(
                    current_executor.execute(request),
                    timeout=request.timeout or self.config.default_timeout,
                )
                
                response.retry_count = retry_count
                response.fallback_used = fallback_used
                response.request_id = request.request_id
                
                if response.success:
                    return response
                
                # Execution failed but no exception
                last_error = response.error or "Unknown error"
                
            except asyncio.TimeoutError:
                last_error = f"Execution timed out after {request.timeout}s"
                logger.warning(f"Timeout executing {request.capability_id}: {last_error}")
                
            except Exception as e:
                last_error = str(e)
                logger.error(f"Error executing {request.capability_id}: {e}")
            
            latency_ms = (time.time() - start_time) * 1000
            
            # Try fallback
            if self.config.fallback_enabled:
                fallback_executors = await self.router.get_fallback_executors(
                    request,
                    current_executor.method,
                    capability_methods,
                )
                
                # Filter out already tried methods
                fallback_executors = [
                    e for e in fallback_executors
                    if e.method not in tried_methods
                ]
                
                if fallback_executors:
                    current_executor = fallback_executors[0]
                    fallback_used = True
                    logger.info(f"Falling back to {current_executor.method.value} for {request.capability_id}")
                    continue
            
            # No fallback, retry with same executor
            retry_count += 1
            if retry_count <= self.config.max_retries:
                delay = self.config.retry_delay * (self.config.retry_backoff ** (retry_count - 1))
                logger.info(f"Retrying {request.capability_id} in {delay}s (attempt {retry_count})")
                await asyncio.sleep(delay)
        
        # All retries exhausted
        return ExecutionResponse(
            success=False,
            error=last_error or "Execution failed after all retries",
            error_code="MAX_RETRIES_EXCEEDED",
            method_used=current_executor.method,
            retry_count=retry_count,
            fallback_used=fallback_used,
            request_id=request.request_id,
        )
    
    async def execute_batch(
        self,
        requests: list[dict[str, Any]],
        concurrency: int = 5,
    ) -> list[ExecutionResponse]:
        """
        Execute multiple requests concurrently.
        
        Args:
            requests: List of request dicts with connector_id, capability_id, params
            concurrency: Maximum concurrent executions
            
        Returns:
            List of responses in same order as requests
        """
        semaphore = asyncio.Semaphore(concurrency)
        
        async def execute_one(req: dict[str, Any]) -> ExecutionResponse:
            async with semaphore:
                return await self.execute(**req)
        
        tasks = [execute_one(req) for req in requests]
        return await asyncio.gather(*tasks)
